### DISCUSSION TITLE
[Flatpak] Move away from debs and RPMs and instead use flatpaks and self updateable appimages.

### DISCUSSION THREAD

**Discussion by xplshn** [minbrowser/min#2258] [Ideas]
What the title says, debs and RPMs aren't needed for things such as a browser, we'd be able to reach more people using flatpak and appimages, for example, in order to install this browser in Void, I had to compile it, not really a user friendly experience

**Comment by 38github** [comment]
+1 since I get errors when installing the .rpm on OpenSUSE Tumbleweed.
`error: unpacking of archive failed on file /usr/share/icons/hicolor/256x256/apps/min.png;659fe83c: cpio: link failed - No data available`

**Comment by NSurtsev** [comment]
We need Flatpak, Snap, AppImage + Windows Portable
Do not know other formats for Mac, but make them too.

**Reply by Tedd-Codes** [reply]
@PalmerAL wait could this work,[Multi Platform Build. Electron]( https://www.electron.build/multi-platform-build.html). I dont know veer much about how Min is built but i know its electron

**Reply by Tedd-Codes** [reply]
Im going to put it on the Snap Store!

**Comment by Tedd-Codes** [comment]
## Cool idea and i might be able to try to add the current version to flat-pack but it would be great!!!

**Comment by xplshn** [comment]
I recently launched a new format called .AppBundle, it works like .AppImages, but if the program you want to bundle is compiled using Musl, you can make it libc-independent, this is great, because you can have a binary that is very small and runs across all Linux-es regardless of the Libc. The only problem is that the binaries aren't mounted with fuse, but rather extracted to a temp dir. https://github.com/xplshn/pelf

**Reply by Tedd-Codes** [reply]
Is there any advantages to using AppBundle and can it be run/installed/executed/whatever without any other software. We want a way to give people a way to run/install without them having to get software to install the installer. Imagine having to get a app to install a app. But if this could work them i would be happy to use your file format because it would be nice to have a new app format. and im happy to support any cool projects and bring together projects.

**Reply by Tedd-Codes** [reply]
Hi sorry, I just looked at the github and it looks like it doesn't not need any thing to run and i think this a good idea. If you could provide more info on how someone would compile to app bundle. Question, does it work like exe on windows.

**Reply by xplshn** [reply]
> Is there any advantages to using AppBundle and can it be run/installed/executed/whatever without any other software. We want a way to give people a way to run/install without them having to get software to install the installer. Imagine having to get a app to install a app. But if this could work them i would be happy to use your file format because it would be nice to have a new app format. and im happy to support any cool projects and bring together projects.

It has no dependencies except SH, and a TAR with Gzip support.
You don't even need to recompile programs, you just use the pelf tool like this:
```
pelf --main-bin /usr/bin/min-browser --output-to ./releases/min-browser-amd64-anyLinux.AppBundle --add-metadata ./min-browser-128x128.png --add-metadata ./min-browser.desktop --add-binary ./some_other_binary_that_is_needed --add-library ./some_library_soft_dependency.so
```

The --add-metadata and --add-* arguments are not necessary. AppBundles ship all the dependencies of a program automatically (as long as those dependencies are visible to the `ldd` command).

**Reply by xplshn** [reply]
> Hi sorry, I just looked at the github and it looks like it doesn't not need any thing to run and i think this a good idea. If you could provide more info on how someone would compile to app bundle. Question, does it work like exe on windows.

I am not sure how .EXEs work. My only experience with Windows is a Windows Xp x64 bits machine I use to game with my dad. Not sure what you mean. .AppBundles can be executed as they are, the user doesn't need to prepare the machine in any way. If the person shipping the appbundle ships the "thumbnailer" inside of the appimage, the appimage will automatically get an icon in the file browser after the first run. Also, if the user has the helper daemon (`pelfd`), it will put all AppBundles in ~/Programs into the user's applications menu.

**Reply by Tedd-Codes** [reply]
Im sorry, i miss-spoke i meant do they act like a exe in how you run it and it will run the app instead of installing it?

**Reply by Tedd-Codes** [reply]
Cool could you show me how someone would install the `thumbnailer` and `helper daemon (pelfd)`? 
So i could make a install script thgat would install those and the Min browser appBundle in one script. i know how to do the install min script part. Apt? Wget?

**Reply by xplshn** [reply]
If you want to add the "thumbnailer" to an AppBundle, you'd do `--add-binary cmd/misc/thumbgen`. thumbgen is at github.com/xplshn/pelf/cmd/misc/thumbgen

And if you want to install pelfd you can do it in many ways:
1. Compile it, clone the repo, go to pelf/cmd/pelfd and run go build, then move the binary to any directory in your $PATH
2. Use `go install github.com/xplshn/pelf/cmd/pelfd@master`
3. Get it from [`bigdl`](https://github.com/xplshn/bigdl), `wget -qO- "https://raw.githubusercontent.com/xplshn/bigdl/master/stubdl" | sh -s -- add pelfd`. This command would add `pelfd` to your `~/.local/bin`

**Reply by xplshn** [reply]
> So i could make a install script thgat would install those and the Min browser appBundle in one script. i know how to do the install min script part. Apt? Wget?

Yes, you can place a wrapper inside of your .AppBundle to handle such tasks. Check out the Examples section

But the idea of AppBundles is to be single-file executables that don't need to be installed.

**Reply by Tedd-Codes** [reply]
so i i wanted to have the appbundle have the icon thing how could i do that

**Reply by Tedd-Codes** [reply]
in a earlier post you said,  The only problem is that the binaries aren't mounted with fuse, but rather extracted to a temp dir
If i use the app bundle thing do i need to do something for the fuse thing?

**Reply by xplshn** [reply]
Nope. Don't worry. Fusermount is not used because not all systems have squashfs-tools installed. This is a disadvantage compared to how AppImages work. But the advantage that this format has, is that you can make libc-independent binaries (if they were compiled from a Musl based distro).

Just read the Readme of github.com/xplshn/pelf if you aren't sure about something. It explains everything in detail.

**Comment by Tedd-Codes** [comment]
@xplshn I think the AppBundle is a awesome idea! could you make a example AppBundle with the optional packages (for icons and installing in programs folder) with the latest version on Min, if so i would like to test it out, im on a raspberry pi.
<details><summary>Min Files</summary>
<p>
<a href="https://objects.githubusercontent.com/github-production-release-asset-2e65be/45319761/84829813-b547-4c59-a9ba-3760d3d53268?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240723%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240723T190953Z&X-Amz-Expires=300&X-Amz-Signature=5ee32ea1b8a5e9c8d7bbbbf315150d87cff7efaa4f2852ce748ac2f845b17eb8&X-Amz-SignedHeaders=host&actor_id=176245405&key_id=0&repo_id=45319761&response-content-disposition=attachment%3B%20filename%3Dmin-1.32.1-arm64.deb&response-content-type=application%2Foctet-stream">Download Min File (for raspberry pi) </a>


<a href="https://github.com/minbrowser/min/releases/tag/v1.32.1"> All Min Download Files (for latest verrsion) </a>
</p>
</details>

**Reply by xplshn** [reply]
I am on Alpine Linux, which uses Musl as its libc, I am not able to run any of the releases. And I am unfamiliar with Node, when I tried to compile Min myself I ended up running out of disk space (for some reason?). Also, I am on x86_64 and don't have a Raspberry. If you are interested, try doing it yourself. It should be simple if you can already run any of the releases. You probably don't even needs scripts of any kind to generate the AppBundle.

You can uncompress the deb files and find out yourself ;)

**Comment by Tedd-Codes** [comment]
@xplshn could you walk me though the step to turn the .deb into a .appbundle. I have installed Pelf. Im not super familiar with pelf and (no offense) your doc is a little confusing.

**Reply by xplshn** [reply]
PELF is the tool used to generate an AppBundle.
1. You must first get decompress the `.deb`
2. Locate the binary files
3. Assemble a directory that contains those files + the files that would be at usr/ and etc,
4. Check if `min` can work with the files being in the same dir as it is or if it REALLY needs those files to be at /usr and so on
5. If `min` does work like that, then you can do the following:
`pelf --main-bin ./path/to/min/binary --output-to ./min-browser.AppBundle --add-metadata ./path/to/min-browser-icon-in-128x128.png --add-metadata ./path/to/min-browser.desktop --add-arbitray ./any/directory/that/min/needs/in/order/to/function`
6. If `min` doesn't work without its data at `/usr` the only thing you can do is check if you recompile it with a modified `$PREFIX`. A `.tar` release would be ideal, since those don't expect to be ran from `/usr/*`. AppImages have the same issue, they can't pack programs which expect their things to be at `/`

Note: The --add-arbitrary flag will add the directory to the ROOT of the AppBundle.
If you want to see the structure of an assembled AppBundle you can do so with `--pbundle_list`:
This is how the output of `mpv.AppBundle --pbundle_list`
![2024-07-23-194550_1231x601_scrot](https://github.com/user-attachments/assets/70ad9d11-5cd1-4f77-8eb7-1723454e6dc7)

NOTE: The examples in the pelf repo are a bit outdated, they use a different option, see their `--pbundle_help`. New bundles generated with the current version of `pelf` will have the `--pbundle_list` option.

**Reply by Tedd-Codes** [reply]
I didnt write min and dont understand what to put in the app bundle, could there be a way to just turn a deb to appbundle, even if it use the files that and needed. Or a way to urn appimage to appbundle

**Reply by xplshn** [reply]
You could add an ./AppImage to an appbundle and then add the libc using: `--add-ld-and-libc`, but, at that point, why? Why not just use an AppImage with a wrapper that does the same?

**Reply by Tedd-Codes** [reply]
I think using the appbundle is a good idea but i still dont know how to, could you use the file attached and send me a appbundle? even if it dosnt work on your system it will work on mine.

**Reply by xplshn** [reply]
> if it dosnt work on your system it will work on mine.

As I've said, you need can only create an AppImage/AppBundle if the program you are trying to package allows having its files in places other than /usr. For that you need to test and adapt `min`, I can't write a wrapper nor an AppRun blindly, I wanted a portable AppImage because I want to run `min` on Alpine Linux, but sadly the developers are only paying attention to Debian, and sometimes, RPM. A Flatpak would be great too, but a portable, MUSL based AppImage would be excellent and cover all distros if done correctly.

**Reply by Tedd-Codes** [reply]
So what file format would be BEST for apline linux

**Reply by xplshn** [reply]
The most portable would be type 2 AppImages that contain all of their dependencies inside + are executed with their ld and libc. (`./ld.so ./bin/program`)

**Reply by Tedd-Codes** [reply]
Ok, i am sorry i cant help you. I know nothing about linux packedges. Maybe someone else can help you. @PalmerAL I think is the creator, and im just here to help add features.
